README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ggplateplot

<!-- badges: start -->

[![R-CMD-check](https://github.com/teunbrand/ggplateplot/workflows/R-CMD-check/badge.svg)](https://github.com/teunbrand/ggplateplot/actions)
[![Codecov test
coverage](https://codecov.io/gh/teunbrand/ggplateplot/branch/master/graph/badge.svg)](https://app.codecov.io/gh/teunbrand/ggplateplot?branch=master)
[![CRAN
status](https://www.r-pkg.org/badges/version/ggplateplot)](https://CRAN.R-project.org/package=ggplateplot)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

Well, well, well… it seems you’ve found this package for plotting:
wells.

The goal of {ggplateplot} is to extend {ggplot2} to make it easier to
make plots resembling cell culture microwell plates. This packages takes
care of some of the layouts, relying on the flexibility of {ggplot2} and
extensions to add layers, scales, theme adjustments et cetera.

## Installation

You can install the development version of {ggplateplot} like so:

``` r
# install.packages("remotes")
remotes::install_github("teunbrand/ggplateplot")
```

## Example

This is a terse example of how one could make a plot from a `matrix` of
values.

``` r
library(ggplot2)
library(ggplateplot)

plate <- matrix(rnorm(24), nrow = 4, ncol = 6)

ggplateplot(plate, aes(fill = value)) +
  geom_well()
```

<img src="man/figures/README-example_matrix-1.png" width="100%" />

Instead of providing a `matrix`, we can also provide long-format data.
It should automatically detect an appropriate layout for the plate based
on the number of rows in the `data` argument, when it has 6, 12, 24, 48,
96 or 384 rows.

``` r
df <- expand.grid(
  x = 1:8,
  y = LETTERS[1:6]
)
df$values <- rnorm(nrow(df))

ggplateplot(df, aes(x, y, fill = values)) +
  geom_well()
```

<img src="man/figures/README-example_long-1.png" width="100%" />

## How does it work?

This package is not a package that instantly makes your plots prettier
or easier. It is a formatting tool for a very particular kind of plot,
of microwell shaped data, in {ggplot2}’s ecosystem.

The `ggplateplot()` function above is a thin wrapper around the
`ggplot()` function. It does a few automated things to make your plots
look more like cell culture plates.

### Geom

First off, the {ggplateplot} package has a specialised layer (geom) for
drawing wells. This isn’t a particularly exciting layer as it just draws
circles as you would with `geom_point()`. A small bit of magic comes
from combining this geom with `coord_plate()`, in that the `size`
aesthetic is synced with the diameter of wells.

``` r
p <- ggplot(df, aes(x, y, fill = values)) +
  geom_well()
p
```

<img src="man/figures/README-geom_well-1.png" width="100%" />

### Coord

The crux of {ggplateplot} is `coord_plate()`, which modifies how several
panel components of a plot are drawn.

#### Theming

The following adjustments to the theming of a plot are made.

-   The `panel.background` and `panel.border` follow the shape of a
    microwell plate. Corners are rounded a little bit and you can ‘bite’
    off some corners with the `corner` argument to resemble a microwell
    plate better.
-   The `panel.grid.major` now draws circles around the wells.
-   The `panel.grid.minor` performs the role of the major panel grid.

``` r
p + coord_plate(corner = "topright", spec = 48) +
  theme(
    panel.grid.major = element_line(colour = "dodgerblue"),
    panel.grid.minor = element_line(colour = "tomato"),
    panel.background = element_rect(colour = "black", fill = "white")
  )
```

<img src="man/figures/README-coord_plate_theme-1.png" width="100%" />

#### Layout

The `coord_plate()` function has a `specs` argument that takes a
description of a well plate. We’ve included descriptions for 6, 12, 24,
48, 96 and 384-well plates in the package. If we mismatch the number of
datapoints with the layout, we might get inappropriate layouts. In the
example below, we’re purposfully giving the wrong layout for the shape
of the data to demonstrate that, among other things, the spacing of
wells is off.

``` r
p + coord_plate(specs = 96)
```

<img src="man/figures/README-coord_plate_spec-1.png" width="100%" />

To offer slightly more flexibility than the standard layouts, you can
set a custom layout with `custom_plate_spec()`.

``` r
df <- data.frame(
  x = rep(1:5, each = 4),
  y = rep(LETTERS[1:4], 5),
  value = rnorm(20)
)

ggplot(df, aes(x, y, fill = value)) +
  geom_well() +
  coord_plate(spec = custom_plate_spec(ncol = 5, nrow = 4))
```

<img src="man/figures/README-coord_plate_custom-1.png" width="100%" />

If you want to even further customise the layout, you can use
`new_plate_spec()` to control even more parameters of the layout.

``` r
spec <- new_plate_spec(
  width = 100, height = 100,
  hor_spacing = 15,
  ver_spacing = 20,
  well_diameter = 12, 
  ncol = 5, nrow = 4,
  corner_size = 25
)

ggplot(df, aes(x, y, fill = value)) +
  geom_well() +
  coord_plate(spec = spec)
```

<img src="man/figures/README-coord_plate_new-1.png" width="100%" />

### Circling back

So what does `ggplateplot()` do exactly? Well, it just tries to
automatically match the shape of your data to an appropriate setting for
`coord_plate()`. If you have data that doesn’t match the dimensions of
one of the standard plates, we recommend that you simply use
`ggplot() + coord_plate()` instead, and customise from there. For
`matrix` input, it additionally converts it to long format and
automatically sets `x` and `y` aesthetics. The values in the matrix are
converted to a `value` column.

``` r
plate <- matrix(rnorm(384), ncol = 24, nrow = 16)

ggplateplot(plate, aes(fill = value)) +
  geom_well()
```

<img src="man/figures/README-round_off-1.png" width="100%" />

## Interacting with ggplot2

While {ggplateplot} makes it easier to make plots of plates, it still
requires some familiarity with the ggplot2 package. The `ggplateplot()`
functions return regular ggplot objects, so they can be combined with
the usual functions one can use to manipulate plots. Below is an example
of combining it with other {ggplot2} functions. We can use an array
instead of matrix input to make variable out of the 3rd dimension, which
we can use to create facets.

``` r
m <- array(1:192, dim = c(8, 12, 2))

ggplateplot(m, aes(fill = value)) +
  geom_point(shape = 24, size = 2.5) +
  scale_fill_viridis_c() +
  scale_y_reverse(name = "Rows", labels = LETTERS[1:8]) +
  facet_wrap(~ Var3) +
  theme(
    panel.background = element_rect(colour = "black", fill = NA),
    panel.grid.major = element_line(colour = "grey90"),
    panel.grid.minor = element_line(colour = "grey97")
  ) +
  labs(x = "Columns")
```

<img src="man/figures/README-complete_example-1.png" width="100%" />


NEWS.md
# ggplateplot 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(ggplateplot)

test_check("ggplateplot")


tests/testthat/test-utils.R
test_that("replace waiver replaces waivers", {
  x <- waiver()
  expect_equal(replace_waiver(x, 2), 2)
  expect_equal(replace_waiver(1, 2), 1)
})

test_that("proect_zero_length returns NULL when length is zero", {
  expect_equal(protect_zero_length(1:2), 1:2)
  expect_null(protect_zero_length(numeric(), NULL))
})


tests/testthat/test-geom_well.R

test_that("vanilla geom_well adopts size from scales", {

  p <- ggplot(mapping = aes(x = 1, y = 1)) +
    geom_well()

  test <- p +
    scale_x_continuous(expand = c(0, 0), limits = c(0, 2)) +
    scale_y_continuous(expand = c(0, 0), limits = c(0, 2))

  l <- layer_grob(test)[[1]]

  expect_equal(as.numeric(l$r), 0.25)

  test <- p +
    scale_x_continuous(expand = c(0, 0), limits = c(-1, 3)) +
    scale_y_continuous(expand = c(0, 0), limits = c(-1, 3))

  l <- layer_grob(test)[[1]]

  expect_equal(as.numeric(l$r), 0.125)

  test <- p +
    scale_x_continuous(expand = c(0, 0), limits = c(0, 2)) +
    scale_y_continuous(expand = c(0, 0), limits = c(-1, 3))

  l <- layer_grob(test)[[1]]

  expect_equal(as.numeric(l$r), 0.125)

  test <- p +
    scale_x_continuous(expand = c(0, 0), limits = c(-1, 3)) +
    scale_y_continuous(expand = c(0, 0), limits = c(0, 2))

  l <- layer_grob(test)[[1]]

  expect_equal(as.numeric(l$r), 0.125)
})

test_that("geom_well adapts in plate context", {
  spec <- new_plate_spec(
    width = 100, height = 100,
    well_spacing = 20, well_diameter = 20,
    ncol = 4, nrow = 4
  )

  df <- expand.grid(x = 1:4, y = 1:4)

  p <- ggplot(df, aes(x, y)) +
    geom_well()

  l <- layer_grob(p + coord_plate(spec = spec))[[1]]

  expect_equal(as.numeric(l$r), rep(0.1, 16))

  spec$well_diameter <- 10

  l <- layer_grob(p + coord_plate(spec = spec))[[1]]

  expect_equal(as.numeric(l$r), rep(0.05, 16))
})


R/geom_well.R

# User function -----------------------------------------------------------

#' Wells
#'
#' This geoms draws circles in a fixed size.
#'
#' @inheritParams ggplot2::geom_point
#'
#' @details The sized is fixed such that a `size = 1` corresponds to 1 unit
#'   on the x- or y-axis, whichever one has the larger range. Alternatively,
#'   when combined with `coord_plate()`, setting `size = 1` is equivalent to
#'   setting the size to a well's diameter.
#'
#' @return A `Layer` object that can be added to a plot.
#' @export
#'
#' @examples
#' df <- expand.grid(LETTERS[1:3], 1:4)
#' df$value <- seq_len(nrow(df))
#'
#' # A basic plot that just draws circles
#' p <- ggplot(df, aes(Var2, Var1, fill = value)) +
#'   geom_well()
#'
#' # With plate coordinates, plot resembles microwell plate
#' p + coord_plate(specs = 12)
geom_well <- function(
  mapping  = NULL,
  data     = NULL,
  stat     = "identity",
  position = "identity",
  ...,
  na.rm       = FALSE,
  show.legend = NA,
  inherit.aes = TRUE
) {
  layer(
    data     = data,
    mapping  = mapping,
    stat     = stat,
    geom     = GeomWell,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params   = list(
      na.rm = na.rm,
      ...
    )
  )
}

# Keys --------------------------------------------------------------------

draw_key_well <- function(data, params, size) {

  circleGrob(
    x = 0.5, y = 0.5,
    r = 0.25 * data$size %||% 1,
    gp = gpar(
      col  = alpha(data$colour %||% "black", data$alpha),
      fill = alpha(data$fill   %||% "black", data$alpha),
      lwd  = data$stroke   %||% 0.5,
      lty  = data$linetype %||% 1
    )
  )
}


# Class -------------------------------------------------------------------

GeomWell <- ggproto(
  NULL, Geom,

  # Aesthetic settings
  required_aes    = c("x", "y"),
  non_missing_aes = c("size"),
  default_aes     = aes(
    size  = 1,  colour = NA, fill = "black",
    alpha = NA, stroke = 0.5, linetype = 1
  ),

  # Drawing function
  draw_panel = function(data, panel_params, coord, na.rm = FALSE) {

    coords <- coord$transform(data, panel_params)
    stroke_size <- coords$stroke
    stroke_size[is.na(stroke_size)] <- 0

    size <- coords$size

    # If coord has specifications, adopt size based on well diameter
    if (!is.null(coord$specs)) {
      mindim <- with(coord$specs, min(width, height))
      size <- 0.5 * size * coord$specs$well_diameter / mindim
    } else {

      # Otherwise, take 1 unit distance on widest axis as size
      dx <- 1 / diff(panel_params$x$continuous_range)
      dy <- 1 / diff(panel_params$y$continuous_range)
      size <- pmin(size * dx, size * dy) / 2
    }

    # Create graphical object
    grob <- circleGrob(
      x  = coords$x,
      y  = coords$y,
      r  = size,
      gp = gpar(
        col  = alpha(coords$colour, coords$alpha),
        fill = alpha(coords$fill,   coords$alpha),
        lwd  = stroke_size,
        lty  = coords$linetype
      )
    )

    # Rename grob
    grob$name <- grobName(grob, "well")
    grob
  },

  draw_key = draw_key_well
)




R/ggplateplot.R
# Generic -----------------------------------------------------------------

#' Plate plots
#'
#' This function is a wrapper for `ggplot()`, that tries to automatically guess
#' an appropriate shape for a microwell layout based on the data.
#'
#' @inheritParams ggplot2::ggplot
#'
#' @return A `ggplot` object.
#' @export
#'
#' @details The 'guessing' part happens by looking at the number of observations
#'   if the `data` argument is a `data.frame`, and the first two dimensions if
#'   the `data` argument is a `matrix` or `array`. If it does not guess the
#'   appropriate shape correctly, it might be easier to use
#'   `ggplot() + ... + coord_shape(spec = ...)` directly.
#'
#'   If the `data` argument is an `array` or `matrix`, this is converted to a
#'   long format `data.frame` and the `x` and `y` aesthetics are populated by
#'   the first two dimensions of the `data` argument.
#'
#' @examples
#' # With long-format data
#' df <- expand.grid(LETTERS[1:2], 1:3)
#' df$value <- seq_len(nrow(df))
#'
#' ggplateplot(df, aes(Var2, Var1, fill = factor(value))) +
#'   geom_well()
#'
#' # With matrices/arrays, cell values become the 'value' column
#' m <- matrix(1:12, nrow = 3, ncol = 4)
#'
#' ggplateplot(m, aes(fill = factor(value))) +
#'   geom_well()
ggplateplot <- function(
  data    = NULL,
  mapping = aes(),
  ...,
  environment = parent.frame()
) {
  UseMethod("ggplateplot", data)
}

# Methods -----------------------------------------------------------------

#' @export
ggplateplot.default <- function(
  data    = NULL,
  mapping = aes(),
  ...,
  environment = parent.frame()
) {
  p <- ggplot(data = data, mapping = mapping, ..., environment = environment)
  if (!is.null(dim(data))) {
    d <- dim(data)
    p <- p + coord_plate(specs = d[1], default = TRUE)
  }
  p
}

#' @export
ggplateplot.array <- function(
  data = NULL,
  mapping = aes(),
  ...,
  environment = parent.frame()
) {
  dim  <- dim(data)
  data <- melt_array(data)

  if (is.null(mapping$x)) {
    mapping$x <- new_quosure(parse_expr(colnames(data)[2]))
  }
  if (is.null(mapping$y)) {
    mapping$y <- new_quosure(parse_expr(colnames(data)[1]))
  }

  p <- ggplot(data = data, mapping = mapping, ..., environment = environment)
  p <- p + coord_plate(specs = prod(dim[1:2]), default = TRUE)
  p

}

# Helpers -----------------------------------------------------------------

melt_array <- function(data) {

  dim <- dim(data)
  nms <- dimnames(data)

  # Get indices
  indices <- lapply(seq_along(dim), function(margin) {
    as.vector(slice.index(data, margin))
  })

  # Lookup dimnames with indices when appropriate
  if (length(nms)) {
    is_named <- lengths(nms) > 0
    indices[is_named] <- Map(
      factor,
      x      = indices[is_named],
      labels = nms[is_named]
    )
  }

  # Give indices variable names
  varnames <- names(nms) %||% rep("", length(dim))
  is_empty <- which(!nzchar(varnames) | varnames == "value")
  varnames[is_empty] <- paste0("Var", is_empty)
  names(indices) <- varnames

  # Convert to data.frame
  ans <- list2DF(indices)
  ans$value <- as.vector(data)

  return(ans)
}


R/specs_plates.R
# Constructor -------------------------------------------------------------

#' Define a new plate layout
#'
#' The `new_plate_spec()` function is used to define the specification of a
#' plate layout. The `custom_plate_spec()` is a wrapper around
#' `new_plate_spec()` that heuristically sets dimensions for quick and dirty
#' layouts.
#'
#' @param width Width of the plate in millimetres.
#' @param height Height of the plate in millimetres.
#' @param hor_spacing Horizontal spacing between wells in millimetres. If
#'   `NULL` (default), the `well_spacing` argument will be used instead.
#' @param ver_spacing Vertical spacing between wells in millimetres. If `NULL`
#'   (default), the `well_spacing` argument will be used instead.
#' @param well_spacing Spacing between wells in millimetres. Useful when
#'   horizontal and vertical spacing is identical.
#' @param well_diameter Diameter of a well in millimetres.
#' @param corner_size Size of the adjacent/opposite sides of the triangle, in
#'   millimeters, when plate has a corner cut off.
#' @param ncol The number of wells in horizontal direction.
#' @param nrow The number of wells in vertical direction.
#'
#' @return A `list` with dimensions that can be understood by the `specs`
#'   argument in `coord_plate()`.
#' @export
#'
#' @examples
#' # Precisely detail a new layout
#' rotated_96well <- new_plate_spec(
#'   width         = 85.5,
#'   height        = 127.8,
#'   well_spacing  = 9,
#'   well_diameter = 6.8,
#'   ncol          = 8,
#'   nrow          = 12
#' )
#'
#' ## Setup dummy data
#' df <- expand.grid(row = 1:12, col = LETTERS[1:8])
#' df$value <- seq_len(nrow(df))
#'
#' ## Use new spec in plot
#' ggplot(df, aes(col, row, fill = value)) +
#'   geom_well() +
#'   coord_plate(spec = rotated_96well)
#'
#' # Quickly sketch a new layout
#' square_2x2 <- custom_plate_spec(ncol = 2, nrow = 2, width = 100, height = 100)
#'
#' ## Setup dummy data
#' df <- expand.grid(row = 1:2, col = LETTERS[1:2])
#' df$value <- seq_len(nrow(df))
#'
#' ## Use spec in plot
#' ggplot(df, aes(col, row, fill = value)) +
#'   geom_well() +
#'   coord_plate(spec = square_2x2)
new_plate_spec <- function(
  width         = 127.8,
  height        = 85.5,
  hor_spacing   = NULL,
  ver_spacing   = NULL,
  well_spacing  = 9,
  well_diameter = 6.8,
  corner_size   = 8,
  ncol          = 12L,
  nrow          = 8L
) {
  # Well dimensions
  hor_spacing  <- hor_spacing %||% well_spacing
  ver_spacing  <- ver_spacing %||% well_spacing
  well_spacing <- min(hor_spacing, ver_spacing)

  # Padding
  pad_width  <- (width  - (ncol - 1) * hor_spacing) / 2
  pad_height <- (height - (nrow - 1) * ver_spacing) / 2

  list(
    width         = width,
    height        = height,
    pad_width     = pad_width,
    pad_height    = pad_height,
    hor_spacing   = hor_spacing,
    ver_spacing   = ver_spacing,
    well_spacing  = well_spacing,
    well_diameter = well_diameter,
    corner_size   = corner_size,
    ncol          = ncol,
    nrow          = nrow
  )
}

#' @export
#' @rdname new_plate_spec
custom_plate_spec <- function(
  ncol = 12L,
  nrow = 8L,
  width  = 127.8,
  height = 85.5
) {

  well_width  <- (0.8 * width) / ncol
  well_height <- (0.8 * height) / nrow
  diameter    <- min(well_width, well_height) * 0.9

  new_plate_spec(
    width  = width,
    height = height,
    hor_spacing = well_width,
    ver_spacing = well_height,
    well_diameter = diameter,
    corner_size = 0.1 * min(width, height),
    ncol = ncol,
    nrow = nrow
  )
}

# Specifications ----------------------------------------------------------

spec_6well <- new_plate_spec(
  width         = 127.8,
  height        = 85.5,
  hor_spacing   = 40,
  ver_spacing   = 38,
  well_diameter = 35,
  ncol          = 3,
  nrow          = 2
)

spec_12well <- new_plate_spec(
  width         = 127.8,
  height        = 85.5,
  well_spacing  = 26,
  well_diameter = 22.8,
  ncol          = 4,
  nrow          = 3
)

spec_24well <- new_plate_spec(
  width         = 127.8,
  height        = 85.5,
  well_spacing  = 19,
  well_diameter = 16.5,
  ncol          = 6,
  nrow          = 4
)

spec_48well <- new_plate_spec(
  width         = 127.8,
  height        = 85.5,
  well_spacing  = 13,
  well_diameter = 10.7,
  ncol          = 8,
  nrow          = 6
)

spec_96well <- new_plate_spec(
  width         = 127.8,
  height        = 85.5,
  well_spacing  = 9,
  well_diameter = 6.8,
  ncol          = 12,
  nrow          = 8
)

spec_384well <- new_plate_spec(
  width         = 127.8,
  height        = 85.5,
  well_spacing  = 4.5,
  well_diameter = 3.65,
  ncol          = 24,
  nrow          = 16
)

all_specs <- list(
  "6"   = spec_6well,
  "12"  = spec_12well,
  "24"  = spec_24well,
  "48"  = spec_48well,
  "96"  = spec_96well,
  "384" = spec_384well
)

# Choosing ----------------------------------------------------------------

choose_wellplate <- function(n) {
  if (n %in% c(6, 12, 24, 48, 96, 384)) {
    name <- paste0("spec_", n, "well")
    obj  <- get(name, envir = asNamespace("ggplateplot"), mode = "list")
    return(obj)
  } else {
    rlang::abort(paste0(
      "A layout specification for ", n, " observation(s) ",
      "has not been implemented."
    ))
  }
}


R/coord_plate.R
# User function -----------------------------------------------------------

#' Plate coordinates
#'
#' This coordinate functions makes your plot look more like a microwell plate.
#'
#' @inheritParams ggplot2::coord_cartesian
#' @param corner The place to draw the notched corner of a plate. Can be one of
#'   `"left"`, `"right"`, `"top"`, `"bottom"`, `"topleft"`,
#'   `"topright"`, `"bottomleft"`, `"bottomright"`, `"none"` or `"all"`.
#' @param specs Either a `list` with plate dimensions or a `numeric(1)` with
#'   one of the common formats, e.g. `6`, `12`, `24`, `48`, `96` or `384` wells.
#'
#' @return A `Coord` object that can be added to a plot.
#' @export
#'
#' @details The `panel.background` theme element now draws a polygon instead
#'  of a rectangle in case `corner != "none"`. Moreover, the `panel.grid.major`
#'  element is used to draw circles at well positions and `panel.grid.minor` is
#'  used to draw regular grid lines. Actual minor breaks are ignored.
#'
#'  The dimensions of most standard plates are roughly based on technical data
#'  sheets available on the Eppendorf website, a manufacturer of cell culture
#'  consumables.
#'
#' @examples
#' # Make basic plot
#' df <- data.frame(x = 1:4, y = 1:4)
#'
#' p <- ggplot(df, aes(x = x, y = y)) +
#'   geom_well()
#'
#' # Using spec and limits to change layout
#' p + coord_plate(spec = 96, xlim = c(1, 12), ylim = c(1, 8))
#' p + coord_plate(spec = 48, xlim = c(1, 8), ylim = c(1, 6))
#'
#' # Changing where the corner is placed
#' p + coord_plate(spec = 24, xlim = c(1, 6), ylim = c(1, 4),
#'                 corner = "left")
coord_plate <- function(
    xlim    = NULL,
    ylim    = NULL,
    expand  = TRUE,
    default = FALSE,
    clip    = "off",
    corner  = "right",
    specs   = 96
) {
  if (!is.list(specs)) {
    specs <- choose_wellplate(specs)
  }
  ratio <- specs$ver_spacing / specs$hor_spacing

  ggproto(
    NULL,
    CoordPlate,
    limits  = list(x = xlim, y = ylim),
    ratio   = ratio,
    expand  = expand,
    clip    = clip,
    corner  = corner,
    specs   = specs,
    default = default
  )
}

# Class -------------------------------------------------------------------

CoordPlate <- ggproto(
  NULL, CoordFixed,

  render_fg = function(self, panel_params, theme) {
    specs  <- self$specs
    corner <- self$corner
    plate_background(
      theme, "panel.border",
      corner, specs$corner_size / specs$height
    )
    # element_render(theme, "panel.border")
  },

  render_bg = function(self, panel_params, theme) {
    guide_wells(
      theme,
      panel_params$x$break_positions_minor(),
      panel_params$x$break_positions(),
      panel_params$y$break_positions_minor(),
      panel_params$y$break_positions(),
      specs  = self$specs,
      corner = self$corner
    )
  },

  modify_scales = function(self, scales_x, scales_y) {
    specs <- self$specs
    x_expand <- expansion(mult = with(specs, pad_width / (width - 2 * pad_width)))
    y_expand <- expansion(mult = with(specs, pad_height / (height - 2 * pad_height)))

    lapply(scales_x, function(sc) {
      sc$expand <- replace_waiver(sc$expand, x_expand)
      if (!sc$is_discrete()) {
        sc$breaks <- replace_waiver(sc$breaks, seq_len(specs$ncol))
      }
      invisible()
    })
    lapply(scales_y, function(sc) {
      sc$expand <- replace_waiver(sc$expand, y_expand)
      if (!sc$is_discrete()) {
        sc$breaks <- replace_waiver(sc$breaks, seq_len(specs$nrow))
      } else {
        sc$limits <- sc$limits %||% rev
      }
      invisible()
    })
    invisible()
  }
)

# Guides ------------------------------------------------------------------

guide_wells <- function(
  theme,
  x.minor,
  x.major,
  y.minor,
  y.major,
  specs,
  corner = "right"
) {

  background <- plate_background(
    theme, "panel.background",
    corner, specs$corner_size / specs$height
  )

  xlines     <- element_render(
    theme, "panel.grid.minor.x",
    x = rep(x.major, each = 2),
    y = rep(0:1, length(x.major)),
    id.lengths = rep(2, length(x.major))
  )
  ylines     <- element_render(
    theme, "panel.grid.minor.y",
    x = rep(0:1, length(y.major)),
    y = rep(y.major, each = 2),
    id.lengths = rep(2, length(y.major))
  )

  well_pos <- expand.grid(x = x.major, y = y.major)
  mindim   <- with(specs, min(height, width))
  radius   <- 0.5 * (specs$well_spacing / mindim)
  elem     <- calc_element("panel.grid.major", theme)

  if (inherits(elem, "element_blank")) {
    wells <- zeroGrob()
  } else {
    wells <- circleGrob(
      x = well_pos$x,
      y = well_pos$y,
      r = radius,
      gp = gpar(
        col  = elem$colour,
        fill = NA,
        lwd  = protect_zero_length(elem$size * .pt),
        lty  = elem$linetype
      )
    )
  }

  ans <- grobTree(
    background,
    xlines,
    ylines,
    wells
  )
  ans$name <- grobName(ans, "grill")
  ans

}

plate_background <- function(theme, element, corner, corner_size) {
  elem <- calc_element(element, theme)
  if (inherits(elem, "element_blank")) {
    return(zeroGrob())
  }

  quarter_circle  <- seq(0, 0.5 * pi, length.out = 9)
  rounded_x <- unit((3 - cos(quarter_circle) * 3) / 85.5, "snpc")
  rounded_y <- unit((3 - sin(quarter_circle) * 3) / 85.5, "snpc")

  fwd <- unit(c(corner_size, 0), "snpc")
  rev <- unit(c(0, corner_size), "snpc")

  if (corner %in% c("topright", "right", "top", "all")) {
    x <- unit(c(1, 1), "npc") - fwd
    y <- unit(c(1, 1), "npc") - rev
  } else {
    x <- unit(1, "npc") - rounded_y
    y <- unit(1, "npc") - rounded_x
  }

  if (corner %in% c("bottomright", "right", "bottom", "all")) {
    nx <- unit(c(1, 1), "npc") - rev
    ny <- unit(c(0, 0), "npc") + fwd
  } else {
    nx <- unit(1, "npc") - rounded_x
    ny <- unit(0, "npc") + rounded_y
  }
  x <- unit.c(x, nx)
  y <- unit.c(y, ny)

  if (corner %in% c("bottomleft", "left", "bottom", "all")) {
    nx <- unit(c(0, 0), "npc") + fwd
    ny <- unit(c(0, 0), "npc") + rev
  } else {
    nx <- unit(0, "npc") + rounded_y
    ny <- unit(0, "npc") + rounded_x
  }
  x <- unit.c(x, nx)
  y <- unit.c(y, ny)

  if (corner %in% c("topleft", "left", "top", "all")) {
    nx <- unit(c(0, 0), "npc") + rev
    ny <- unit(c(1, 1), "npc") - fwd
  } else {
    nx <- unit(0, "npc") + rounded_x
    ny <- unit(1, "npc") - rounded_y
  }
  x <- unit.c(x, nx)
  y <- unit.c(y, ny)

  gp <- gpar(
    lwd  = protect_zero_length(elem$size * .pt),
    col  = elem$colour,
    fill = elem$fill,
    lty  = elem$linetype
  )

  pathGrob(
    x = x, y = y,
    gp = gp
  )
}
